Fennec migration: Fenix crashes at start

## Preconditions:
Have Firefox Release installed and logged in.
## Steps to reproduce
1. Replace Fennec with Fenix using a build from https://github.com/mozilla-mobile/fenix/wiki/Fennec-Migration;
2. Launch Fenix.
### Expected behavior
Fenix opens.
### Actual behavior
Fenix crashes.
### Notes:
[logcat-crash.txt](https://github.com/mozilla-mobile/android-components/files/3840189/logcat-crash.txt)
[Video](https://drive.google.com/file/d/1ERAMHz0FlvcAbjgXUoOEF19udlHCZI4s/view)
### Device information

* Android device: Nexus 6P (Android 8.1.0) - ARM; Samsung Galaxy S8 (Android 9) - ARM; Google Pixel 2 (Android 9) - ARM64.
* Fenix version: Nightly 191109 GV 71

There's definitely another startup crash here, which I can reproduce locally - and we've seen it in Sentry. I will use this bug as a placeholder for that crash.

```
                         E  java.lang.RuntimeException: Unable to create application org.mozilla.fenix.MigratingFenixApplication: java.lang.IllegalStateException: Unrecognized migration type: Mig
                            ration$FxA
                         E      at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5743)
                         E      at android.app.ActivityThread.-wrap1(Unknown Source:0)
                         E      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1656)
                         E      at android.os.Handler.dispatchMessage(Handler.java:106)
                         E      at android.os.Looper.loop(Looper.java:164)
                         E      at android.app.ActivityThread.main(ActivityThread.java:6494)
                         E      at java.lang.reflect.Method.invoke(Native Method)
                         E      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
                         E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
                         E  Caused by: java.lang.IllegalStateException: Unrecognized migration type: Migration$FxA
                         E      at mozilla.components.support.migration.MigrationResultsStore.fromJSON(MigrationResultsStore.kt:12)
                         E      at mozilla.components.support.base.utils.SharedPreferencesCache.getCached(SharedPreferencesCache.kt:2)
                         E      at mozilla.components.support.migration.FennecMigrator.migrateAsync(FennecMigrator.kt:3)
                         E      at org.mozilla.fenix.MigratingFenixApplication.setupApplication(MigratingFenixApplication.kt:144)
                         E      at org.mozilla.fenix.MigratingFenixApplication.onCreate(MigratingFenixApplication.kt:2)
                         E      at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1119)
                         E      at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5740)
                         E      ... 8 more
    mozac/CrashReporter  I  Received crash: UncaughtExceptionCrash(throwable=java.lang.RuntimeException: Unable to create application org.mozilla.fenix.MigratingFenixApplication: java.lang.Illega
                            lStateException: Unrecognized migration type: Migration$FxA, breadcrumbs=[])
```

--- a/components/support/migration/src/main/java/mozilla/components/support/migration/MigrationResultsStore.kt
+++ b/components/support/migration/src/main/java/mozilla/components/support/migration/MigrationResultsStore.kt
@@ -57,6 +57,7 @@ internal class MigrationResultsStore(context: Context) : SharedPreferencesCache<
                 Migration.Bookmarks.javaClass.simpleName -> Migration.Bookmarks
                 Migration.OpenTabs.javaClass.simpleName -> Migration.OpenTabs
                 Migration.Gecko.javaClass.simpleName -> Migration.Gecko
+                Migration.FxA.javaClass.simpleName -> Migration.FxA
                 else -> throw IllegalStateException("Unrecognized migration type: $migrationName")
             }
             result[migration] = MigrationRun(version = migrationVersion, success = migrationSuccess)

--- a/components/support/migration/src/test/java/mozilla/components/support/migration/FennecMigratorTest.kt
+++ b/components/support/migration/src/test/java/mozilla/components/support/migration/FennecMigratorTest.kt
@@ -30,6 +30,7 @@ import java.lang.IllegalStateException
 import kotlinx.coroutines.CompletableDeferred
 import mozilla.components.service.fxa.sharing.ShareableAccount
 import mozilla.components.support.test.argumentCaptor
+import org.mockito.Mockito.reset
 
 @RunWith(AndroidJUnit4::class)
 class FennecMigratorTest {
@@ -422,6 +423,13 @@ class FennecMigratorTest {
         }
 
         verifyZeroInteractions(accountManager)
+
+        // Does not run FxA migration again.
+        with(migrator.migrateAsync().await()) {
+            assertEquals(0, this.size)
+        }
+
+        verifyZeroInteractions(accountManager)
     }
 
     @Test
@@ -438,6 +446,13 @@ class FennecMigratorTest {
             assertEquals(1, this.size)
             assertTrue(this.containsKey(Migration.FxA))
             assertTrue(this.getValue(Migration.FxA).success)
+        }
+
+        verifyZeroInteractions(accountManager)
+
+        // Does not run FxA migration again.
+        with(migrator.migrateAsync().await()) {
+            assertEquals(0, this.size)
         }
 
         verifyZeroInteractions(accountManager)
@@ -469,6 +484,14 @@ class FennecMigratorTest {
         assertEquals("252fsvj8932vj32movj97325hjfksdhfjstrg23yurt267r23", captor.value.authInfo.kSync)
         assertEquals("0b3ba79bfxdf32f3of32jowef7987f", captor.value.authInfo.kXCS)
         assertEquals("fjsdkfksf3e8f32f23f832fwf32jf89o327u2843gj23", captor.value.authInfo.sessionToken)
+
+        // Does not run FxA migration again.
+        reset(accountManager)
+        with(migrator.migrateAsync().await()) {
+            assertEquals(0, this.size)
+        }
+
+        verifyZeroInteractions(accountManager)
     }
 
     @Test
@@ -498,5 +521,13 @@ class FennecMigratorTest {
         assertEquals("252bc4ccc3a239fsdfsdf32fg32wf3w4e3472d41d1a204890", captor.value.authInfo.kSync)
         assertEquals("0b3ba79b18bd9fsdfsdf4g234adedd87", captor.value.authInfo.kXCS)
         assertEquals("fsdfjsdffsdf342f23g3ogou97328uo23ij", captor.value.authInfo.sessionToken)
+
+        // Does not run FxA migration again.
+        reset(accountManager)
+        with(migrator.migrateAsync().await()) {
+            assertEquals(0, this.size)
+        }
+
+        verifyZeroInteractions(accountManager)
     }
 }

